libafl/src/observers/stacktrace.rs

```diff
@@ -0,0 +1,319 @@
+        }
+    }
+}
+
+/// An enum encoding the types of harnesses
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub enum HarnessType {
+    /// Harness type when the harness is rust code
+    RUST,
+    /// Harness type when the harness is linked via FFI (e.g C code)
+    FFI,
+}
+
+/// An observer looking at the backtrace after the harness crashes
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct BacktraceObserver {
+    observer_name: String,
+    harness_type: HarnessType,
+    hash: Option<u64>,
+}
+
+impl BacktraceObserver {
+    /// Creates a new [`StacktraceObserver`] with the given name.
+    #[must_use]
+    pub fn new(observer_name: &str, harness_type: HarnessType) -> Self {
+        match harness_type {
+            HarnessType::RUST => stacktrace_hooks::setup_rust_panic_hook(),
+            HarnessType::FFI => unsafe { stacktrace_hooks::setup_signal_handler() },
+        }
+        Self {
+            observer_name: observer_name.to_string(),
+            harness_type,
+            hash: None,
+        }
+    }
+
+    /// Setup the shared memory and store it in [`BACKTRACE_HASH_VALUE`]
+    pub fn setup_shmem() {
+        let shmem_provider = StdShMemProvider::new();
+        println!("panic hook is being set");
+        let shmem = shmem_provider.unwrap().new_map(5000).unwrap();
+        unsafe {
+            BACKTRACE_HASH_VALUE = BacktraceHashValueWrapper::Shmem(shmem);
+        }
+    }
+
+    /// Init the [`BACKTRACE_HASH_VALUE`] to BacktraceHashValueWrapper::StaticVariable(0)
+    pub fn setup_static_variable() {
+        unsafe {
+            BACKTRACE_HASH_VALUE = BacktraceHashValueWrapper::StaticVariable(0);
+        }
+    }
+}
+
+impl ObserverWithHashField for BacktraceObserver {
+    /// Gets the hash value of this observer.
+    #[must_use]
+    fn hash(&self) -> &Option<u64> {
+        &self.hash
+    }
+
+    /// Updates the hash value of this observer.
+    fn update_hash(&mut self, hash: u64) {
+        self.hash = Some(hash);
+    }
+
+    /// Clears the current hash value
+    fn clear_hash(&mut self) {
+        self.hash = None;
+    }
+}
+
+impl Default for BacktraceObserver {
+    fn default() -> Self {
+        Self::new("BacktraceObserver", HarnessType::RUST)
+    }
+}
+
+impl<I, S> Observer<I, S> for BacktraceObserver
+where
+    I: Debug,
+{
+    fn pre_exec(&mut self, _state: &mut S, _input: &I) -> Result<(), Error> {
+        Ok(())
+    }
+
+    fn post_exec(&mut self, _state: &mut S, _input: &I) -> Result<(), Error> {
+        unsafe {
+            let hash = BACKTRACE_HASH_VALUE.get_stacktrace_hash();
+            println!("hash from parent process is {}", hash);
+            self.update_hash(hash);
+        }
+        Ok(())
+    }
+}
+
+impl Named for BacktraceObserver {
+    fn name(&self) -> &str {
+        &self.observer_name
+    }
+}
+
+/// returns the recommended ASAN runtime flags to capture the backtrace correctly
+pub fn get_asan_runtime_flags() -> String {
```

Comment:
Our recommended flags for asan (maybe change the name of the function to default or recommended) are these https://github.com/AFLplusplus/LibAFL/blob/main/libafl_targets/src/common.c